feat(User): Basic user preferences, GET User by attribute.

binh-dam-ibigroup opened the pull request:
### Checklist

- [X] Appropriate branch selected _(all PRs must first be merged to `dev` before they can be merged to `master`)_
- [x] Any modified or new methods or classes have helpful JavaDoc and code is thoroughly commented
- [na] The description lists all applicable issues this PR seeks to resolve
- [na] The description lists any configuration setting(s) that differ from the default settings
- [x] All tests and CI builds passing

### Description

This PR adds the following:
* An api route to get entities by attribute, used in the new MOD UI to retrieve user data using an auth0 id. (The Mongo ID is not stored or communicated by Auth0).
* User fields and types related to creating user accounts on middleware.

binh-dam-ibigroup committed:
feat(User): Add basic user preferences for account creation.

binh-dam-ibigroup committed:
feat(user): Add spark route to get user by attribute value.

binh-dam-ibigroup committed:
docs: Update method/field documentation.

binh-dam-ibigroup committed:
improvement(User): Add other persisted user fields.

binh-dam-ibigroup committed:
fix(http OPTIONS): Add OPTIONS method to /find/{..}/{..} path.

br648 commented on src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java:
```diff
@@ -90,6 +89,19 @@ public void bind(final SparkSwagger restApi) {
+            // Get one entity by field.
+            .get(path(ROOT_ROUTE + FIND_PATH)
+                    .withDescription("Returns a '" + classToLowercase + "' entity whose field has the specified value, or 404 if not found.")
+                    .withPathParam().withName("attribute").withDescription("The attribute the entity to search.").and()
```
"The attribute the entity to search." This doesn't make sense.

br648 commented on src/main/java/org/opentripplanner/middleware/models/User.java:
```diff
@@ -11,16 +13,54 @@
+    public String notificationChannel;
+
+    /** Phone number for SMS notifications. */
+    public String phoneNumber;
```
Does the User class now contain all required fields? If so, perhaps remove the TODO (on line 16).

br648 commented on src/main/java/org/opentripplanner/middleware/models/UserNotificationChannel.java:
```diff
@@ -0,0 +1,9 @@
+ * Describes the means to notify a user:
+ * none (no notification), via email, via sms).
+ */
+public enum UserNotificationChannel {
```
If we decide that enums can't be saved to Mongo this will need to be changed.

br648 commented on src/main/java/org/opentripplanner/middleware/persistence/TypedPersistence.java:
```diff
@@ -122,6 +122,16 @@ public T getById (String id) {
         return mongoCollection.find(eq(id)).first();
     }
 
+    /**
+     * Retrieves the first document that matches an field value, or null if there is no match.
+     * @param field The field to scan.
+     * @param value The value to find.
+     * @return A document whose specified field has the specified value.
+     */
```
This is the only method that has this amount of detail in the comment (@param and @return). I'm not sure what the policy is on this, but probably easier to remove to be inline with the other comments?

binh-dam-ibigroup replied:
Method is redundant and I went ahead and removed it.

br648 reviewed (commented):
Some minor comments. All good otherwise.

binh-dam-ibigroup committed:
improvement(User): Make User.notificationChannels a string list. Address PR comments.

binh-dam-ibigroup commented:
Note: last commit will convert `notificationChannel` into `notificationChannels` to allow notifications with both email and sms.

evansiroky commented on src/main/java/org/opentripplanner/middleware/persistence/TypedPersistence.java:
```diff
@@ -122,6 +122,16 @@ public T getById (String id) {
+     * @param value The value to find.
+     * @return A document whose specified field has the specified value.
+     */
+    public T getByField (String field, String value) {
```
I would like this to be named `getFirstByFieldValue` to be clear that it is returning a single record and getting the first result whatever that may be.

binh-dam-ibigroup replied:
I think this method will become `getOneFiltered` once #14 is merged.

binh-dam-ibigroup replied:
Method is redundant. I went ahead and removed it.

evansiroky commented on src/main/java/org/opentripplanner/middleware/models/UserLocation.java:
```diff
@@ -0,0 +1,18 @@
+    public String icon;
+
+    /** The type/purpose of this location (free form) (e.g. home, work, shopping...) */
+    public String type;
```
Needs `lat`, `lon` added.

evansiroky commented on src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java:
```diff
@@ -188,6 +209,22 @@ private T getObjectForId(Request req, String id) {
+    /**
+     * Convenience method for extracting the attribute/field param from the HTTP request.
+     */
+    private T getObjectWithField(Request req, String field, String value) {
```
Would like this renamed to `getFirstObjectByFieldValue` to indicate that it is getting the first record that matches a field value.

evansiroky commented on src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java:
```diff
@@ -139,6 +151,15 @@ private T getOne(Request req, Response res) {
+    /**
+     * HTTP endpoint to get one entity specified by a field value.
+     */
+    private T getOneByField(Request req, Response res) {
```
I'd like this renamed to `getFirstByFieldValue` to indicate that this is getting the first record that matches the given field value.

evansiroky reviewed (changes requested):
See comments about renaming things to indicate that certain methods are returning the first item matching a field value. I feel this is important because the first result from a search on a field without distinct values could end up returning inconsistent records.

evansiroky commented on src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java:
```diff
@@ -90,6 +89,19 @@ public void bind(final SparkSwagger restApi) {
             )
 
+            // Get one entity by field.
+            .get(path(ROOT_ROUTE + FIND_PATH)
```
Each of these get API endpoints don't have any security associated with them. Any requester could just send over an arbitrary Auth0 user id and it would return the user object for another user.

binh-dam-ibigroup replied:
@evansiroky This is addressed by #21 in https://github.com/ibi-group/otp-middleware/blob/b827e04b81d9b2fd31a7b558ab9f3b0cd846a473/src/main/java/org/opentripplanner/middleware/auth/Auth0Utils.java#L15 so we *will* be making those checks.

binh-dam-ibigroup replied:
Actually, it is also handled by #13 in https://github.com/ibi-group/otp-middleware/blob/94dcefa65446530ba5854afb442aac7d416c5312/src/main/java/org/opentripplanner/middleware/auth/Auth0Connection.java#L38

evansiroky replied:
Ok, I'll overlook this for now since #21 has needed checks.

evansiroky reviewed (changes requested):
Just discovered a big security risk, so I'm marking this PR as "requesting changes" until all get methods have authorization checks added to ensure data is returned only when proper authorization can be confirmed.

Also, it appears that the deletion methods need proper authorization checks as well.

binh-dam-ibigroup committed:
revert(user): Revert notificationChannel to String.

binh-dam-ibigroup commented:
Caution: On parent branch `add-sub-users`, a similar "filtered query" was introduced. Need to use it. See https://github.com/ibi-group/otp-middleware/blob/77b9e500c1f308690e41bec6cbd893567aaf3667/src/main/java/org/opentripplanner/middleware/auth/Auth0UserProfile.java#L43

binh-dam-ibigroup commented:
> Just discovered a big security risk, so I'm marking this PR as "requesting changes" until all get methods have authorization checks added to ensure data is returned only when proper authorization can be confirmed.
> 
> Also, it appears that the deletion methods need proper authorization checks as well.

Suggestion: address the quoted comment in #13 or #14. Thoughts? @evansiroky and @landonreed

binh-dam-ibigroup committed:
refactor(ApiController): Reuse Persistence.getOneFiltered()

binh-dam-ibigroup committed:
refactor(ApiController): Rename method.

binh-dam-ibigroup committed:
improvement(User): Add lat, lon, name to user saved locations.

evansiroky commented on src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java:
```diff
@@ -197,29 +198,26 @@ private String deleteOne(Request req, Response res) {
+    /**
+     * Log if an object was not found for a request.
+     */
+    private T checkAndLogIfNull(Request req, T object, String field, String value) {
```
This is too much abstraction to save just a little bit of repeated code. It makes things less readable, so I think the original way of doing this in `getObjectForId` is preferred.

evansiroky replied:
Or maybe a hybrid approach would be better where you can change this method to be something like:

```java
private void haltIfNull(T object, Request req, String message) {
    if (object == null) {
        logMessageAndHalt(req, HttpStatus.NOT_FOUND_404, message);
    }
}
```

evansiroky reviewed (changes requested):
See https://github.com/ibi-group/otp-middleware/pull/15/files#r426110256 and https://github.com/ibi-group/otp-middleware/pull/15/files#r426108133 and https://github.com/ibi-group/otp-middleware/pull/15/files#r428252282.

binh-dam-ibigroup committed:
fix(Auth0User): Cache token expiration time.

binh-dam-ibigroup committed:
refactor(tests): Remove dependency on configuration.

binh-dam-ibigroup committed:
refactor: Reinline method extracted for this work.

binh-dam-ibigroup committed:
refactor(ApiController): Address PR comments

landonreed committed:
refactor(auth): refactor auth0 token expiration handling

binh-dam-ibigroup committed:
feat(Auth0UserController): Add Auth0UserController to get users by auth0 token.

landonreed committed:
test(auth0): fix Auth0 test, update test config

binh-dam-ibigroup committed:
refactor(ApiController): Remove methods related to getting entity by attribute value.

BREAKING CHANGE: API route /find/{attribute}/{value} is removed.

binh-dam-ibigroup committed:
Merge pull request #20 from ibi-group/check-token-expiration-ltr

Check token expiration changes

binh-dam-ibigroup committed:
refactor(Auth0Connection): Remove unused method.

binh-dam-ibigroup committed:
Merge pull request #19 from ibi-group/check-token-expiration

Check token expiration

binh-dam-ibigroup committed:
Merge branch 'basic-user-prefs' into get-user-by-auth0-token

binh-dam-ibigroup committed:
refactor(UserController): Add fromtoken route to UserController.

binh-dam-ibigroup committed:
test(UserControllerTest): Add UserControllerTest

binh-dam-ibigroup committed:
test(UserControllerTest): Add test with null profile.

binh-dam-ibigroup committed:
Merge branch 'user-sub-types' into basic-user-prefs

binh-dam-ibigroup committed:
fix(Merge): Fix merge issues.

binh-dam-ibigroup committed:
refactor(OtpUser): Cleanup imports.

binh-dam-ibigroup committed:
Merge branch 'basic-user-prefs' into get-user-by-auth0-token

binh-dam-ibigroup committed:
refactor(UserController): Address PR comments.

binh-dam-ibigroup committed:
refactor(UserControllerTest): Make test pass. Tweak method comments.

binh-dam-ibigroup committed:
refactor(UserController): Address more PR comments.

binh-dam-ibigroup committed:
refactor(UserController): Use getUserFromRequest from Auth0Connection.

binh-dam-ibigroup committed:
docs(UserController): Tweak comment.

binh-dam-ibigroup committed:
Merge branch 'user-sub-types' into basic-user-prefs

binh-dam-ibigroup committed:
Merge branch 'add-auth0' into user-sub-types

binh-dam-ibigroup committed:
Merge branch 'user-sub-types' into basic-user-prefs

binh-dam-ibigroup committed:
Merge branch 'basic-user-prefs' into get-user-by-auth0-token

evansiroky commented on src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java:
```diff
@@ -90,6 +92,19 @@ public void bind(final SparkSwagger restApi) {
+                    .withPathParam().withName("value").withDescription("The desired value for the specified attribute.").and()
+                    // .withResponses(...) // FIXME: not implemented (requires source change).
+                    .withResponseType(clazz),
+                    this::getOneByField, JsonUtils::toJson
```
`getOneByField` is undefined

binh-dam-ibigroup replied:
Thanks for pointing it out!

evansiroky reviewed (changes requested):
Travis builds failing probably due to undefined method call. Probably good to go once that is fixed.

binh-dam-ibigroup committed:
docs(ApiController): Fix comment typo.

binh-dam-ibigroup committed:
fix(ApiController): Reinstate deleted method while resolving merge conflicts.

binh-dam-ibigroup committed:
style(UserController, ApiController): Fix spacing, fix typo.

br648 reviewed (dismissed):
Looks good!

binh-dam-ibigroup committed:
Merge branch 'dev' into basic-user-prefs

binh-dam-ibigroup committed:
Merge branch 'basic-user-prefs' into get-user-by-auth0-token

binh-dam-ibigroup committed:
refactor(ApiController): Remove unused method.

br648 reviewed (dismissed):
All good and unti test passes.

binh-dam-ibigroup committed:
docs(OtpUserController): Fix log and API description.

binh-dam-ibigroup committed:
Merge pull request #22 from ibi-group/get-user-by-auth0-token

Get user by auth0 token

evansiroky commented on src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java:
```diff
@@ -274,10 +297,17 @@ private T createOrUpdate(Request req, Response res) {
+    /**
+     * Get entity attribute value from request.
+     */
+    private String getParamFromRequest(Request req, String paramName) {
```
Can we rename this to `getRequiredParamFromRequest` and add a note that this method will halt the request if the param is not provided?

evansiroky commented on src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java:
```diff
@@ -58,22 +59,44 @@ public ApiController (String apiPrefix, TypedPersistence<T> persistence, String
+     * a request with /path/subpath will be handled by the method for /path/subpath.
+     * @param baseEndPoint The end point to which to add the methods.
+     */
+    protected void buildEndPoint(ApiEndpoint baseEndPoint) {
```
Please rename `EndPoint` to `Endpoint` to maintain consistency.

evansiroky commented on src/main/java/org/opentripplanner/middleware/controllers/api/OtpUserController.java:
```diff
@@ -16,10 +23,35 @@
     }
 
+    @Override
+    protected void buildEndPoint(ApiEndpoint baseEndPoint) {
```
Please rename `EndPoint` to `Endpoint` to maintain consistency.

evansiroky commented on src/main/java/org/opentripplanner/middleware/controllers/api/OtpUserController.java:
```diff
@@ -49,4 +81,18 @@ boolean preDeleteHook(OtpUser user, Request req) {
+        OtpUser user = profile.otpUser;
+
+        if (user == null) {
+            logMessageAndHalt(req, HttpStatus.NOT_FOUND_404, String.format(NO_USER_WITH_AUTH0_ID_MESSAGE, profile.auth0UserId), null);
```
This should return an HTTP 403 error since it seems that whatever request this is will fail if an authenticated user is not present in the request.

binh-dam-ibigroup replied:
@evansiroky Please describe a situation where this would be the case.

For users who just created an Auth0 login and have an `Auth0UserProfile`,
but have not completed the account setup form yet, the `OtpUser` profile would naturally be 404 not found until they complete the account setup form and you wouldn't want to 403 forbid that user to try to retrieve their account.

evansiroky replied:
Huh, I guess this one could go either way.

evansiroky commented on src/test/java/org/opentripplanner/middleware/auth/Auth0UsersTest.java:
```diff
@@ -0,0 +1,43 @@
+        // Verify that token is not expired soon after it set.
+        assertFalse(Auth0Users.getCachedToken().isStale());
+        // Verify that seconds until expiration has decreased since first set.
+        assertTrue(Auth0Users.getCachedToken().secondsUntilExpiration() < TOKEN_DURATION_SECONDS);
```
Since this is dependent on the time, I recommend putting in a Thead.sleep or something to be extra sure that some amount of time has elapsed.

evansiroky reviewed (changes requested):
See most recent comments.

br648 reviewed (dismissed):
Looks good.

binh-dam-ibigroup committed:
refactor: Rename methods, clarify comments, remove unused imports.

BREAKING CHANGE: ApiController.bindEndPoint has been renamed to ApiController.bindEndpoint

landonreed commented on src/main/java/org/opentripplanner/middleware/models/OtpUser.java:
```diff
@@ -1,10 +1,45 @@
+     * Some identity services, such as Auth0, manage the email verification process and status,
+     * so this field is only for reference.
+     */
+    public boolean isEmailVerified;
```
@binh-dam-ibigroup, is this field populated from the Auth0 field? We're not allowing users to create an account if their email has not already been verified, correct? This suggests to me that this field is no longer required.

binh-dam-ibigroup replied:
I will remove. We don't use this field and get that information straight from Auth0.

binh-dam-ibigroup replied:
Same for recent locations.

landonreed commented on src/main/java/org/opentripplanner/middleware/models/OtpUser.java:
```diff
@@ -1,10 +1,45 @@
+    public boolean isEmailVerified;
+
+    /** Whether the phone number has been verified. */
+    public boolean isPhoneNumberVerified;
```
Please add TODO because we have not yet added code to verify the phone number.

landonreed commented on src/main/java/org/opentripplanner/middleware/models/OtpUser.java:
```diff
@@ -1,10 +1,45 @@
+    /**
+     * Notification preference for this user
+     * ("email", "sms", or "none").
+     * NOTE: This could become an enum, e.g. https://jira.mongodb.org/browse/JAVA-268.
```
This comment is very cryptic. Please just bump the mongo version if that resolves the enum storage issue.

binh-dam-ibigroup replied:
Updating the comment. The Mongo driver supports the desired functionality, however there is additional setup needed.

landonreed reviewed (changes requested):
Looks good, just a few changes.

br648 commented on src/main/java/org/opentripplanner/middleware/controllers/api/OtpUserController.java:
```diff
@@ -16,10 +23,35 @@
     }
 
+    @Override
+    protected void buildEndpoint(ApiEndpoint baseEndPoint) {
```
I think Evan mentioned this, lowercase 'point' so baseEndpoint.

binh-dam-ibigroup replied:
Fixed. Thanks.

br648 reviewed (dismissed):
One very minor item, apart from that all good!

binh-dam-ibigroup committed:
refactor(ApiController): Change case of other endpoints.

binh-dam-ibigroup committed:
refactor(OtpUser): Remove unused fields, add TODOs.

br648 reviewed (approved):
Looks good!

landonreed reviewed (approved):
Good stuff. Thanks for the changes.
